### DISCUSSION TITLE
Can you return both data and errors for the initial request?

### DISCUSSION THREAD

**Discussion by rajansolanki** [TanStack/query#3176] [Q&A] [answered]
With graphql, it's possible for a response to contain both data and errors [at the same time](https://github.com/graphql/graphql-spec/blob/main/spec/Section%207%20--%20Response.md). Right now, if a background refetch of a query fails, that query then goes into a state where both its `data` and `error` are populated. Is it possible to recreate that behaviour for an initial request?

Apollo has [something similar](https://www.apollographql.com/docs/react/data/error-handling/#graphql-error-policies) if you set the error policy of a query to `all`

**Comment by TkDodo** [comment, marked as answer]
I don't see how its currently possible, because react-query needs a resolved or rejected promise to go into success or error state. You could do it in user-land by:
- returning a resolved promise
- inspect `data` and move the related errors to a separate `error` field that you can then return from your custom hook.

but react-query will not go into error state then, so there won't be retries for example.

**Reply by rajansolanki** [reply]
Thanks @TkDodo! In that case I think I'll throw a custom `Error` which will accept `data`, that way the query correctly goes into an error state, but the data is still accessible.

Something like this:

```ts
const json = await res.json();
if (!json.errors) return json.data;

throw new CustomError(json.errors, json.data);
```

```ts
const useData = () => {
  const { data, error } = useQuery();

  return data || error.data;
};
```
